module/state_synchronization/execution_data_service.go

```diff
@@ -0,0 +1,406 @@
+	bcw, br := blobs.IncomingBlobChannel(s.maxBlobSize)
+
+	done := make(chan struct{})
+	var serializeErr error
+
+	// start serialization goroutine
+	go func() {
+		defer close(done)
+		defer bcw.Close()
+
+		if serializeErr = s.serializer.Serialize(bcw, v); serializeErr != nil {
+			logger.Debug().Err(serializeErr).Msg("failed to serialize execution data")
+
+			return
+		}
+
+		serializeErr = bcw.Flush()
+
+		if serializeErr != nil {
+			logger.Debug().Err(serializeErr).Msg("flushed execution data")
+		}
+	}()
+
+	cids, storeErr := s.storeBlobs(ctx, br, logger)
+
+	<-done
+
+	if storeErr != nil {
+		// the blob channel may be closed as a result of the context being canceled,
+		// in which case we should return the context error.
+		if errors.Is(storeErr, blobs.ErrClosedBlobChannel) && ctx.Err() != nil {
+			return nil, ctx.Err()
+		}
+
+		return nil, storeErr
+	}
+
+	return cids, serializeErr
+}
+
+// Add constructs a blob tree for the given ExecutionData and adds it to the blobservice, and then returns the root CID.
+func (s *executionDataServiceImpl) Add(ctx context.Context, sd *ExecutionData) (cid.Cid, error) {
+	logger := s.logger.With().Str("block_id", sd.BlockID.String()).Logger()
+	logger.Debug().Msg("adding execution data")
+
+	s.metrics.ExecutionDataAddStarted()
+
+	start := time.Now()
+	cids, err := s.addBlobs(ctx, sd, logger)
+
+	if err != nil {
+		s.metrics.ExecutionDataAddFinished(time.Since(start), false, 0)
+
+		return cid.Undef, fmt.Errorf("failed to add execution data blobs: %w", err)
+	}
+
+	var blobTreeNodes int
+
+	for {
+		cidArr := zerolog.Arr()
+
+		for _, cid := range cids {
+			cidArr.Str(cid.String())
+		}
+
+		logger.Debug().Array("cids", cidArr).Msg("added blobs")
+
+		blobTreeNodes += len(cids)
+
+		if len(cids) == 1 {
+			s.metrics.ExecutionDataAddFinished(time.Since(start), true, blobTreeNodes)
+
+			return cids[0], nil
+		}
+
+		if cids, err = s.addBlobs(ctx, cids, logger); err != nil {
+			s.metrics.ExecutionDataAddFinished(time.Since(start), false, blobTreeNodes)
+
+			return cid.Undef, fmt.Errorf("failed to add cid blobs: %w", err)
+		}
+	}
+}
+
+// retrieveBlobs retrieves the blobs for the given CIDs from the blobservice, and sends them to the given BlobSender
+// in the order specified by the CIDs.
+func (s *executionDataServiceImpl) retrieveBlobs(parent context.Context, bs *blobs.BlobSender, cids []cid.Cid, logger zerolog.Logger) error {
+	defer bs.Close()
+
+	ctx, cancel := context.WithCancel(parent)
+	defer cancel()
+
+	blobChan := s.blobService.GetBlobs(ctx, cids) // initiate a batch request for the given CIDs
+	cachedBlobs := make(map[cid.Cid]blobs.Blob)
+	cidCounts := make(map[cid.Cid]int) // used to account for duplicate CIDs
+
+	for _, c := range cids {
+		cidCounts[c] += 1
+	}
+
+	for _, c := range cids {
+		blob, ok := cachedBlobs[c]
+
+		if !ok {
+			var err error
+
+			blob, err = s.findBlob(blobChan, c, cachedBlobs, logger)
```

Comment:
the call to `findBlob` will be blocking, is there any guarantee that it will return within certain duration?